views/bigquery.go

```diff
@@ -0,0 +1,300 @@
+	TimeStep    string
+	Filter      QueryFilter
+	BreakdownBy []string
+	Detailed    bool
+}
+
+// GetSummaryQueryArgs returns the args to be used for a views summary query if
+// the provided query can be replaced by a summary query.
+// i.e. it is querying for the aggregate metrics of a single playback ID.
+func (s QuerySpec) GetSummaryQueryArgs() (playbackID string, ok bool) {
+	playbackID = s.Filter.PlaybackID
+	pidFilter := QueryFilter{PlaybackID: playbackID, UserID: s.Filter.UserID}
+
+	ok = playbackID != "" && !s.Detailed && s.Filter == pidFilter &&
+		s.From == nil && s.To == nil && s.TimeStep == "" &&
+		len(s.BreakdownBy) == 0
+
+	if !ok {
+		playbackID = ""
+	}
+	return
+}
+
+var viewershipBreakdownFields = map[string]string{
+	"device_type":    "device_type",
+	"device":         "device",
+	"cpu":            "cpu",
+	"os":             "os",
+	"browser":        "browser",
+	"browser_engine": "browser_engine",
+	"continent":      "playback_continent_name",
+	"country":        "playback_country_name",
+	"subdivision":    "playback_subdivisions_name",
+	"timezone":       "playback_timezone",
+	"viewer_id":      "viewer_id",
+}
+
+var allowedTimeSteps = map[string]bool{
+	"hour":  true,
+	"day":   true,
+	"week":  true,
+	"month": true,
+	"year":  true,
+}
+
+func buildViewsEventsQuery(table string, spec QuerySpec) (string, []interface{}, error) {
+	query := squirrel.Select(
+		"countif(play_intent) as view_count",
+		"sum(playtime_ms) / 60000.0 as playtime_mins").
+		From(table).
+		Where("account_id = ?", spec.Filter.UserID).
+		Limit(maxBigQueryRows + 1)
+	query = withPlaybackIdFilter(query, spec.Filter.PlaybackID)
+
+	if spec.Detailed {
+		query = query.Columns(
+			"avg(ttff_ms) as ttff_ms",
+			"avg(rebuffer_ratio) as rebuffer_ratio",
+			"avg(if(error_count > 0, 1, 0)) as error_rate",
+			"avg(if(exit_before_start, 1, 0)) as exits_before_start")
+	}
+
+	if creatorId := spec.Filter.CreatorID; creatorId != "" {
+		query = query.Where("creator_id_type = ?", "unverified")
+		query = query.Where("creator_id = ?", creatorId)
+	}
+
+	if timeStep := spec.TimeStep; timeStep != "" {
+		if !allowedTimeSteps[timeStep] {
+			return "", nil, fmt.Errorf("invalid time step: %s", timeStep)
+		}
+
+		query = query.
+			Columns(fmt.Sprintf("timestamp_trunc(time, %s) as time_interval", timeStep)).
+			GroupBy("time_interval").
+			OrderBy("time_interval")
+	}
+
+	if from := spec.From; from != nil {
+		query = query.Where("time >= timestamp_millis(?)", from.UnixMilli())
+	}
+	if to := spec.To; to != nil {
+		query = query.Where("time < timestamp_millis(?)", to.UnixMilli())
+	}
+
+	for _, by := range spec.BreakdownBy {
+		field, ok := viewershipBreakdownFields[by]
+		if !ok {
+			return "", nil, fmt.Errorf("invalid breakdown field: %s", by)
+		}
+		query = query.Columns(field).GroupBy(field)
+	}
+
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return "", nil, err
+	}
+
+	return sql, args, nil
+}
+
+type ViewershipEventRow struct {
```

Comment:
Would move this struct and the Summary one before the build functions or at the top